[docs] update philosophy.md (finally) (#13808)
huggingface/diffusers: docs/source/en/conceptual/philosophy.md

At Hugging Face, we call this design the **single-file policy** which means that almost all of the code of a certain class should be written in a single, self-contained file. To read more about the philosophy, you can have a look
at [this blog post](https://huggingface.co/blog/transformers-design-philosophy).

In Diffusers, we follow this philosophy for pipelines, schedulers, and models alike. Some older models predate this convention and are kept as-is; all new model architectures live in their own self-contained files. See the [Models](#models) section below for details.

Great, now you should have generally understood why 🧨 Diffusers is designed the way it is 🤗.
We try to apply these design principles consistently across the library. Nevertheless, there are some minor exceptions to the philosophy or some unlucky design choices. If you have feedback regarding the design, we would ❤️  to hear it [directly on GitHub](https://github.com/huggingface/diffusers/issues/new?assignees=&labels=&template=feedback.md&title=).

## Design Philosophy in Details

> Now, let's look a bit into the nitty-gritty details of the design philosophy. Diffusers consists of [models](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models), [schedulers](https://github.com/huggingface/diffusers/tree/main/src/diffusers/schedulers), and two ways to compose them into a runnable workflow: standard [pipelines](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines) (monolithic, one task per pipeline class) and [Modular Diffusers](../modular_diffusers/overview) (composable, block-based). Let's walk through more in-detail design decisions for each.

Reviewer: ```suggestion
Now for the nitty-gritty details of the design philosophy. Diffusers gives you two ways to compose [models](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models) and [schedulers](https://github.com/huggingface/diffusers/tree/main/src/diffusers/schedulers) into a runnable workflow: standard [pipelines](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines), which are monolithic with one task per pipeline class, and Modular Diffusers, which is composable and block-based. The sections below cover pipelines, modular pipelines, models, and schedulers in turn.
```

### Pipelines

> Pipelines are intended **only** for inference. Pipelines are designed to be easy to use (therefore do not follow [*Simple over easy*](#simple-over-easy) 100%) — they should be readable, self-explanatory, easy to tweak, and loosely seen as examples of how to use [models](#models) and [schedulers](#schedulers). Pipelines are not feature complete, to build feature-complete user interfaces on top of Diffusers, consider using [Modular Diffusers](../modular_diffusers/overview).

Reviewer 2: ```suggestion
Pipelines, either standard or modular, are intended **only** for inference. Pipelines are designed to be easy to use (therefore do not follow [*Simple over easy*](#simple-over-easy) 100%) — they should be readable, self-explanatory, easy to tweak, and loosely seen as examples of how to use [models](#models) and [schedulers](#schedulers). Pipelines are not feature complete, to build feature-complete user interfaces on top of Diffusers, consider using [Modular Diffusers](../modular_diffusers/overview).
```

> Pipelines are intended **only** for inference. Pipelines are designed to be easy to use (therefore do not follow [*Simple over easy*](#simple-over-easy) 100%) — they should be readable, self-explanatory, easy to tweak, and loosely seen as examples of how to use [models](#models) and [schedulers](#schedulers). Pipelines are not feature complete, to build feature-complete user interfaces on top of Diffusers, consider using [Modular Diffusers](../modular_diffusers/overview).

Reviewer: ```suggestion
Pipelines, standard or modular, are intended only for inference. They're designed to be easy to use (so they don't follow [Simple over easy](#simple-over-easy) 100%): readable, self-explanatory, easy to tweak, and best seen as examples of how to use [models](#models) and [schedulers](#schedulers). They aren't feature complete. To build feature-complete user interfaces on top of Diffusers, use [Modular Diffusers](../modular_diffusers/overview).
```

- Pipelines all inherit from [`DiffusionPipeline`].
- Every pipeline consists of different model and scheduler components, that are documented in the [`model_index.json` file](https://huggingface.co/stable-diffusion-v1-5/stable-diffusion-v1-5/blob/main/model_index.json), are accessible under the same name as attributes of the pipeline and can be shared between pipelines with [`DiffusionPipeline.components`](https://huggingface.co/docs/diffusers/main/en/api/diffusion_pipeline#diffusers.DiffusionPipeline.components) function.
- Every pipeline should be loadable via the [`DiffusionPipeline.from_pretrained`](https://huggingface.co/docs/diffusers/main/en/api/diffusion_pipeline#diffusers.DiffusionPipeline.from_pretrained) function.
- Every pipeline should have one and only one way to run it via a `__call__` method. The naming of the `__call__` arguments should be consistent across all pipelines.
- Pipelines should be named after the task they are intended to solve.

> ### Modular Pipelines

Reviewer: ```suggestion
### Modular Diffusers
```

> Modular Pipelines is the composable alternative to standard pipelines: workflows are built from reusable **pipeline blocks** that can be mixed, matched, swapped, and shared. Where standard pipelines are loose reference examples of how to use models and schedulers, Modular Pipelines is the recommended path for building feature-complete user interfaces on top of Diffusers, and for the community to build and share new pipelines in a decentralized way.

Reviewer: ```suggestion
Modular Diffusers is the composable alternative to standard pipelines. You build a workflow from reusable *pipeline blocks* that you can mix, match, swap, and share. Standard pipelines are loose reference examples of how to use models and schedulers. Modular Diffusers is the recommended path for building feature-complete user interfaces on top of Diffusers, and for the community to build and share new pipelines in a decentralized way.
```

The following design principles are followed:

> - Modular pipelines also follow the single-file policy. Each modular pipeline lives in its own folder under [`src/diffusers/modular_pipelines/`](https://github.com/huggingface/diffusers/tree/main/src/diffusers/modular_pipelines). A modular pipeline folder contains multiple files corresponding to each stage of the workflow — `encoders.py`, `before_denoise.py`, `denoise.py`, `decoders.py`, `modular_blocks_<model>.py` for assembly, and `modular_pipeline.py` for the per-model `ModularPipeline` subclass. Modular pipelines don't cross-import each other.

Reviewer: ```suggestion
- Modular pipelines follow the single-file policy. Each one lives in its own folder under [src/diffusers/modular_pipelines/](https://github.com/huggingface/diffusers/tree/main/src/diffusers/modular_pipelines), and the folder splits the workflow across one file per stage: `encoders.py`, `before_denoise.py`, `denoise.py`, and `decoders.py`. Two more files complete the pipeline. `modular_blocks_<model>.py` assembles the stages, and `modular_pipeline.py` defines the per-model [`ModularPipeline`] subclass. Modular pipelines don't cross-import each other.
```

> - Each modular pipeline is defined as a set of `ModularPipelineBlocks` — leaf blocks in the stage files (`encoders.py`, `before_denoise.py`, `denoise.py`, `decoders.py`), then assembled into the full workflow in `modular_blocks_<model>.py` using container classes like `SequentialPipelineBlocks` and `AutoPipelineBlocks`. Unlike `DiffusionPipeline`, which both defines the computation logic and runs it, modular splits these into two concepts: blocks are pure definitions (declare inputs, outputs, and component dependencies — they do not hold weights and cannot be executed), and a `ModularPipeline` is the runnable counterpart, created via `.init_pipeline(repo_id)`. Keeping blocks stateless and weight-free is what makes them freely composable, swappable, and shareable across workflows.

Reviewer: ```suggestion
- Each modular pipeline is defined as a set of [`ModularPipelineBlocks`]. Leaf blocks live in the stage files (`encoders.py`, `before_denoise.py`, `denoise.py`, `decoders.py`), and `modular_blocks_<model>.py` assembles them into the full workflow with container classes like [`SequentialPipelineBlocks`] and [`AutoPipelineBlocks`]. This splits apart two concepts that [`DiffusionPipeline combines`]. A block is a pure definition. It declares inputs, outputs, and component dependencies, but holds no weights and can't run. A [`ModularPipeline`], created with `.init_pipeline(repo_id)`, is the runnable counterpart. Keeping blocks stateless and weight-free is what makes them freely composable, swappable, and shareable across workflows.
```

> - To support a new workflow/task, write the task-specific block(s), compose them with existing ones, and declare the new workflow in `_workflow_map` on the top-level block assembly. Unlike `DiffusionPipeline`, each `ModularPipeline` can support many workflows (text2img, img2img, inpaint, etc.).

Reviewer: ```suggestion
- To support a new task, write the task-specific blocks, compose them with existing ones, and register the workflow in `_workflow_map` on the top-level block assembly. A single `ModularPipeline` can support many workflows, such as text-to-image, image-to-image, and inpainting, whereas a `DiffusionPipeline` runs only one.
```

See the [Modular Diffusers documentation](../modular_diffusers/overview) for the full design and usage guide.

### Models

> Models are designed as configurable toolboxes that are natural extensions of [PyTorch's Module class](https://pytorch.org/docs/stable/generated/torch.nn.Module.html). They should follow the **single-file policy**. Some older models predate this convention and are kept as-is as legacy exceptions — not patterns to follow for new models. E.g. the original [`UNet2DConditionModel`] class was used for different UNet variations.

Reviewer: ```suggestion
Models are designed as configurable toolboxes that are natural extensions of [PyTorch's Module class](https://pytorch.org/docs/stable/generated/torch.nn.Module.html). They should follow the *single-file policy*. Some older models predate this convention and are kept as-is. Treat them as legacy exceptions, not patterns to follow for new models. For example, the original [`UNet2DConditionModel`] class was used for several UNet variations.
```

The following design principles are followed:

> - Each model architecture type lives in its own folder under [`src/diffusers/models`](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models) (e.g. [`transformers/`](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/transformers), [`autoencoders/`](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/autoencoders), [`unets/`](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/unets)), and each model family has its own file within that folder (e.g. [`transformer_flux.py`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_flux.py), [`transformer_wan.py`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_wan.py)).

Reviewer: ```suggestion
- Each model architecture type lives in its own folder under [src/diffusers/models](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models), for example [transformers/](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/transformers), [autoencoders/](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/autoencoders), or [unets/](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/unets). Each model family has its own file within that folder, such as [transformer_flux.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_flux.py) and [transformer_wan.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_wan.py).
```

> - Models follow the **single-file policy** — each model file should be self-contained, except for a small number of standard modules that every model uses identically (e.g. timestep embeddings, normalization layers), which can be imported from [`embeddings.py`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py) and [`normalization.py`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/normalization.py).

Reviewer: ```suggestion
- Models follow the single-file policy. Each model file should be self-contained, except for a small number of standard modules that every model uses identically, such as timestep embeddings and normalization layers. Import those from [embeddings.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py) and [normalization.py](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/normalization.py).
```
